sale_rental/models/sale_order.py

```diff
@@ -0,0 +1,300 @@
+# Copyright 2014-2016 Akretion (http://www.akretion.com)
+# @author Alexis de Lattre <alexis.delattre@akretion.com>
+# Copyright 2016 Sodexis (http://sodexis.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api, _
+from odoo.exceptions import ValidationError, UserError
+from odoo.tools import float_compare
+from dateutil.relativedelta import relativedelta
+from odoo.tools import DEFAULT_SERVER_DATETIME_FORMAT,\
+    DEFAULT_SERVER_DATE_FORMAT
+from datetime import datetime
+import odoo.addons.decimal_precision as dp
+import logging
+
+
+logger = logging.getLogger(__name__)
+# TODO : block if we sell a rented product already sold => state
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    @api.multi
+    def action_confirm(self):
+        res = super(SaleOrder, self).action_confirm()
+        for order in self:
+            for line in order.order_line:
+                if line.rental_type == 'new_rental':
+                    self.env['sale.rental'].create(line._prepare_rental())
+                    # Check availability of picking moves
+                    # and changing picking status also
+                    line.move_ids._action_assign()
+                elif line.rental_type == 'rental_extension':
+                    if not line.extension_rental_id or\
+                            not line.extension_rental_id.in_move_id:
+                        continue
+                    line.extension_rental_id.in_move_id.date_expected =\
+                        line.end_date
+                    line.extension_rental_id.in_move_id.date = line.end_date
+                elif line.sell_rental_id:
+                    if line.sell_rental_id.out_move_id.state != 'done':
+                        raise UserError(_(
+                            'Cannot sell the rental %s because it has '
+                            'not been delivered')
+                            % line.sell_rental_id.display_name)
+                    line.sell_rental_id.sell_move_id._action_assign()
+                    line.sell_rental_id.in_move_id._action_cancel()
+        return res
+
+    @api.multi
+    def action_cancel(self):
+        """
+            In case cancelling a SO which sells rental product.
+            Picking in should be created manually
+        """
+        res = super(SaleOrder, self).action_cancel()
+        for order in self:
+            for line in order.order_line:
+                if line.rental_type == 'new_rental':
+                    sale_rental = self.env['sale.rental'].search(
+                        [('start_order_line_id', '=', line.id)], limit=1)
+                    sale_rental.out_move_id._action_cancel()
+                    sale_rental.in_move_id._action_cancel()
```

Comment:
You out_move_id and in_move_id will be cancelled in super function. You don't need to do it again.